### ISSUE TITLE
wxD2D1Factory is created with D2D1_DEBUG_LEVEL_WARNING even in Release builds

### ISSUE THREAD

**Issue by PBfordev** [wxWidgets/wxWidgets#22034]
wxD2D1Factory is created with `D2D1_FACTORY_OPTIONS` set to `D2D1_DEBUG_LEVEL_WARNING` even in Release builds like this (irrelevant parts removed, full code [here](https://github.com/wxWidgets/wxWidgets/blob/4a43ea882781f4be40c6064d4918697008b57353/src/msw/graphicsd2d.cpp#L560)):
``` cpp
D2D1_FACTORY_OPTIONS factoryOptions = {D2D1_DEBUG_LEVEL_NONE};

#if defined(__WXDEBUG__) && defined(__VISUALC__) && wxCHECK_VISUALC_VERSION(11)
    factoryOptions.debugLevel = D2D1_DEBUG_LEVEL_WARNING;
#endif  //__WXDEBUG__
```
`__WXDEBUG__` is by default defined in the wxWidgets library even in Release builds (see [here](https://github.com/wxWidgets/wxWidgets/blob/593864a81686854ce5512159f1071566d073c8f7/include/wx/debug.h#L46)), which means that code using Direct2D APIs will get a `D2D DEBUG ERROR` exception thrown before it gets a chance to process the returned `HRESULT`.

I believe this is not correct and may make the application using `wxD2D1Factory` quite unusable (this is not a catchable C++ exception), see Known issues in #22029.

I think the condition should be changed, to also take into account `NDEBUG`, i.e. to
``` cpp
#if !defined(NDEBUG) && defined(__WXDEBUG__) && defined(__VISUALC__) && wxCHECK_VISUALC_VERSION(11)
```
Is there any reason why not to do that? 

FWIW, [the official Direct2D SVG example](https://docs.microsoft.com/en-us/samples/microsoft/windows-universal-samples/d2dsvgimage/) has it like this ( `_DEBUG` is defined by the MSVC when using the Debug version of the CRT):
``` cpp
D2D1_FACTORY_OPTIONS options;
ZeroMemory(&options, sizeof(options));

#if defined(_DEBUG)
    // If the project is in a debug build, enable Direct2D debugging via SDK Layers.
    options.debugLevel = D2D1_DEBUG_LEVEL_INFORMATION;
#endif
```

**Comment by vadz** [comment]
I don't have any real D2D experience, so I don't know how useful is it to enable the debug level, but I would think that we should do it when actually debugging, i.e. when `wxDISABLE_DEBUG_SUPPORT()` is not used. There is not really straightforward way to check for this, but I think we could use `wxTheAssertHandler` as a proxy, e.g.
```cpp
#ifdef __WXDEBUG__
    if ( wxTheAssertHandler )
        options.debugLevel = D2D1_DEBUG_LEVEL_INFORMATION;
#endif // __WXDEBUG__
```

What do you think?

**Comment by PBfordev** [comment]
I don't have any D2D experience either. Additionally, this may be the first time I am hearing about undocumented `wxTheAssertHandler`, sorry. 

You are much more knowledgeable about wxWidgets than me. 

However, if we really want to get the exceptions only when actually debugging, would it not be better to use `wxIsDebuggerRunning()` instead? If those exceptions are thrown when the program is not running under a debugger, the program just silently quits.

**Comment by vadz** [comment]
`wxTheAssertHandler != nullptr` is the condition which distinguishes debug "builds" of wx application from release ones -- except they're not really "builds" but "runs", as this is a run-time rather than compile-time condition. So this is the most faithful translation of `#ifdef _DEBUG` test.

Now if the only effect of using a debug level is to throw these exceptions, then I agree that using `wxIsDebuggerRunning()` could be even better -- but note that this would mean enabling debug level even in release runs. And perhaps this wouldn't be desirable, e.g. if this slows down the drawing?

Again, I just don't know what does enabling the debug level does exactly in D2D (I skimmed the MSDN description but it doesn't really explain it). Perhaps we could catch these exceptions and translate them to asserts/`wxLogDebug()` calls?

**Comment by PBfordev** [comment]
> Again, I just don't know what does enabling the debug level does exactly in D2D (I skimmed the MSDN description but it doesn't really explain it). Perhaps we could catch these exceptions and translate them to asserts/`wxLogDebug()` calls?

As I wrote, I do not think this a C++ exception, at least I could not catch it with `catch (...)` and catching SEH exception in a class method is not easy. It looks like this
![D2_DEBUG_ERROR_exc](https://user-images.githubusercontent.com/12495521/150560768-393fe933-27dc-4bea-bc14-da6d1775c3e3.png)

Perhaps it is just D2D Debug Layer doing something like
``` cpp
::OutputDebugString("D2 DEBUG ERROR - ...");
::__debugbreak();
```
> Now if the only effect of using a debug level is to throw these exceptions, then I agree that using `wxIsDebuggerRunning()` could be even better -- but note that this would mean enabling debug level even in release runs. And perhaps this wouldn't be desirable, e.g. if this slows down the drawing?

Direct2D Debug Layer documentation states: "The debug layer should never be active in the release version of an application." To me, that would generally imply to never turn D2D debugging on in the release build of wxWidgets regardless of the value of `wxTheAssert`, i.e., rely on `NDEBUG` which is AFAIK always defined in wxWidgets release builds (is it really?).

**Comment by vadz** [comment]
Catching SEH exceptions is not pretty, but not really a problem and we have some `wxSEH_xxx` macros in `wx/msw/seh.h`. But it does look like it just triggers a breakpoint, which can't be caught anyhow when running under a debugger.

But I still think we should use `wxTheAssertHandler` check to decide whether to enable it, the checks done by D2D seem quite similar to our asserts in intention, so IMO it makes sense to enable them if, and only if, asserts are enabled.

More broadly, we followed Microsoft model of separate debug and release builds until wx 2.8, but, familiar as it is, it's really not the best one and the idea that you shouldn't test/debug different code from what is actually used in production is really important -- I've seen multiple problems that only arose in release builds and a non negligible part of them was due to compiling with different `_DEBUG` values. So since wx 3.0 we basically have a single build (we still have 2 when using MSVS because this is just more familiar under Windows, but they only differ in their code generation options, but still compile the same code) and debugging features can be turned on and off dynamically. The _only_ place where we depend on `_DEBUG` (or `NDEBUG` which is the more standard complement to it) is the program initialization, where we enable debugging features by default if `NDEBUG` is not defined.

**TL;DR** Let's enable D2D asserts iff our own asserts are enabled.

**Comment by PBfordev** [comment]
I do not want to argue. 

All I wanted to say was that there is a difference between D2D Debug Break and `wxASSERT`/`assert()`. The former just makes the application suddenly quit (before the code gets the chance to handle the failed API call) with no explanation while the latter shows the message showing what went wrong (at least in MSVC debug build) and let's the user decide what to do.

Also, AFAIK, aside from the Debug Heap, is not there another difference between Debug and Release MSVC wxWidgets builds at least when it comes to standard containers, where the debug version uses checked iterators?

Last but not least, `assert`s are used to check for programmer's errors. However, the D2D Debug throws even for things that are not really under the programmer's control, such as invalid SVG file format.

**Comment by vadz** [comment]
FWIW I see your point(s) and I agree that this is not an exact equivalent of checking for "debug build", but I still believe that it's quite close, semantically, and I'm all but sure that this is the closest thing to it that we have.

I also think that using debug checks for SVG correctness is wrong precisely for the reason you mention, but I guess that in our usage SVGs will come from (semi-)hardcoded resources, so, if you squint enough, this could be seen as programmer's mistake. As I said, not exactly the same, but still closish...
